<a href="https://colab.research.google.com/github/wiz124/chem169-git/blob/main/Li_Harry_RID_017_code.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [61]:
#Exercise 0
import pandas as pd

df=pd.read_excel('mtb_with_localization.xlsx')
display(df)

/usr/local/lib/python3.12/dist-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


Entry    Reviewed  \
0     A0A089QRB9    reviewed   
1         I6WXK4    reviewed   
2         I6WZG6    reviewed   
3         I6X235    reviewed   
4         I6X8D2    reviewed   
...          ...         ...   
3992      V5QPR9  unreviewed   
3993      V5QPS4    reviewed   
3994      V5QQR7  unreviewed   
3995      V5QQS8  unreviewed   
3996      V5QRX2  unreviewed   

                                          Protein names  \
0     Mycolipanoate synthase (EC 2.3.1.252) (Mycocer...   
1     Triple specificity protein phosphatase PtpB (E...   
2     Type 1 encapsulin shell protein (Culture filtr...   
3     ADP-ribose pyrophosphatase (EC 3.6.1.13) (8-ox...   
4                Polyketide synthase Pks13 (EC 2.3.1.-)   
...                                                 ...   
3992        Acid and phagosome regulated protein A AprA   
3993              Putative antitoxin Rv3098B/RVBD_3098B   
3994                                  Conserved protein   
3995                               Probable transposase   
3996        Acid and phagosome regulated protein B AprB   

                         Gene Names  \
0      msl3 pks3 pks4 Rv1180/Rv1181   
1                ptpB mptpB Rv0153c   
2                 enc cfp29 Rv0798c   
3                            Rv1700   
4                     pks13 Rv3800c   
...                             ...   
3992                   aprA Rv2395A   
3993  Rv3098B RVBD_3098B P425_03228   
3994                        Rv1366A   
3995                        Rv3349c   
3996                   aprB Rv2395B   

                                               Organism  Length  Annotation  \
0     Mycobacterium tuberculosis (strain ATCC 25618 ...    2085         5.0   
1     Mycobacterium tuberculosis (strain ATCC 25618 ...     276         5.0   
2     Mycobacterium tuberculosis (strain ATCC 25618 ...     265         5.0   
3     Mycobacterium tuberculosis (strain ATCC 25618 ...     207         5.0   
4     Mycobacterium tuberculosis (strain ATCC 25618 ...    1733         5.0   
...                                                 ...     ...         ...   
3992  Mycobacterium tuberculosis (strain ATCC 25618 ...      71         1.0   
3993  Mycobacterium tuberculosis (strain ATCC 25618 ...      66         1.0   
3994  Mycobacterium tuberculosis (strain ATCC 25618 ...      86         1.0   
3995  Mycobacterium tuberculosis (strain ATCC 25618 ...     246         1.0   
3996  Mycobacterium tuberculosis (strain ATCC 25618 ...      54         1.0   

                                          Function [CC]  \
0     FUNCTION: Polyketide synthase involved in the ...   
1     FUNCTION: Essential virulence factor that prom...   
2     FUNCTION: Shell component of a type 1 encapsul...   
3     FUNCTION: Catalyzes the hydrolysis of ADP-ribo...   
4     FUNCTION: Involved in the biosynthesis of myco...   
...                                                 ...   
3992                                                NaN   
3993  FUNCTION: Putative antitoxin component of a po...   
3994                                                NaN   
3995                                                NaN   
3996                                                NaN   

                              Subcellular location [CC]  \
0     SUBCELLULAR LOCATION: Cell membrane {ECO:00002...   
1     SUBCELLULAR LOCATION: Secreted {ECO:0000269|Pu...   
2     SUBCELLULAR LOCATION: Encapsulin nanocompartme...   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
3992                                                NaN   
3993                                                NaN   
3994                                                NaN   
3995                                                NaN   
3996                                                NaN   

                                               Sequence  
0     MRTATATSVAVIGMACR

In [62]:
#Exercise 1

location=[]
for entry in df['Subcellular location [CC]']:
  if 'membrane' in str(entry).lower():
    location.append('membrane')
  elif 'secreted' in str(entry).lower():
    location.append('secreted')
  elif 'cytoplasm' in str(entry).lower():
    location.append(  'cytoplasm')
  else:
    location.append(None)

df['location']=location
df.dropna(subset=['location'], inplace=True)
display(df)

print(len(df[df['location'] == 'membrane']))
print(len(df[df['location'] == 'cytoplasm']))
print(len(df[df['location'] == 'secreted']))



Entry    Reviewed  \
0     A0A089QRB9    reviewed   
1         I6WXK4    reviewed   
2         I6WZG6    reviewed   
7         I6XFS7    reviewed   
11        I6Y1Q2    reviewed   
...          ...         ...   
3895      Q10883  unreviewed   
3928      Q6MX20  unreviewed   
3934      Q6MX47  unreviewed   
3954      Q79FG5  unreviewed   
3955      Q79FG7  unreviewed   

                                          Protein names  \
0     Mycolipanoate synthase (EC 2.3.1.252) (Mycocer...   
1     Triple specificity protein phosphatase PtpB (E...   
2     Type 1 encapsulin shell protein (Culture filtr...   
7     RNA/DNA methyltransferase Rv2966c (16S rRNA m2...   
11    Manganese-dependent transcriptional repressor ...   
...                                                 ...   
3895                          Possible hydrogenase HycQ   
3928                     histidine kinase (EC 2.7.13.3)   
3934       Probable conserved integral membrane protein   
3954               TVP38/TMEM64 family membrane protein   
3955                Probable conserved membrane protein   

                        Gene Names  \
0     msl3 pks3 pks4 Rv1180/Rv1181   
1               ptpB mptpB Rv0153c   
2                enc cfp29 Rv0798c   
7                          Rv2966c   
11                mntR sirR Rv2788   
...                            ...   
3895                   hycQ Rv0086   
3928                       Rv2998A   
3934                       Rv0318c   
3954                       Rv2306A   
3955                       Rv2219A   

                                               Organism  Length  Annotation  \
0     Mycobacterium tuberculosis (strain ATCC 25618 ...    2085         5.0   
1     Mycobacterium tuberculosis (strain ATCC 25618 ...     276         5.0   
2     Mycobacterium tuberculosis (strain ATCC 25618 ...     265         5.0   
7     Mycobacterium tuberculosis (strain ATCC 25618 ...     188         5.0   
11    Mycobacterium tuberculosis (strain ATCC 25618 ...     228         5.0   
...                                                 ...     ...         ...   
3895  Mycobacterium tuberculosis (strain ATCC 25618 ...     488         1.0   
3928  Mycobacterium tuberculosis (strain ATCC 25618 ...      67         1.0   
3934  Mycobacterium tuberculosis (strain ATCC 25618 ...     264         1.0   
3954  Mycobacterium tuberculosis (strain ATCC 25618 ...     197         1.0   
3955  Mycobacterium tuberculosis (strain ATCC 25618 ...     140         1.0   

                                          Function [CC]  \
0     FUNCTION: Polyketide synthase involved in the ...   
1     FUNCTION: Essential virulence factor that prom...   
2     FUNCTION: Shell component of a type 1 encapsul...   
7     FUNCTION: SAM-dependent methyltransferase that...   
11    FUNCTION: Transcriptional regulator that contr...   
...                                                 ...   
3895                                                NaN   
3928                                                NaN   
3934                                                NaN   
3954                                                NaN   
3955                                                NaN   

                              Subcellular location [CC]  \
0     SUBCELLULAR LOCATION: Cell membrane {ECO:00002...   
1     SUBCELLULAR LOCATION: Secreted {ECO:0000269|Pu...   
2     SUBCELLULAR LOCATION: Encapsulin nanocompartme...   
7     SUBCELLULAR LOCATION: Secreted {ECO:0000269|Pu...   
11       SUBCELLULAR LOCATION: Cytoplasm {ECO:0000305}.   
...                                                 ...   
3895  SUBCELLULAR LOCATION: Cell membrane {ECO:00002...   
3928  SUBCELLULAR LOCATION: Cell membrane {ECO:00002...   
3934  SUBCELLULAR LOCATION: Membrane {ECO:0000256|AR...   
3954  SUBCELLULAR LOCATION: Cell membrane {ECO:00002...   
3955  SUBCELLULAR LOCATION: Cell membrane {ECO:00002...   

                                               Sequence   location  
0     MRTATATSVAVIGMACRL

641
315
136


In [63]:
#Exercise 2
import numpy as np

length=[]
for sequence in df['Sequence']:
  length.append(len(sequence))

df['length']=length

print(df['length'].max())
print(df['length'].min())
print(df['length'].mean())

x = df[['length']].to_numpy()
y = df[['location']].to_numpy()



2108
50
378.0100732600733


In [64]:
#Exercise 3
def count_amino_acids(sequence):
    amino_acids = list('ACDEFGHIKLMNPQRSTVWY')
    sequence=str(sequence)
    return {aa: sequence.count(aa) for aa in amino_acids}

amino_acids = list('ACDEFGHIKLMNPQRSTVWY')

aa_counts = pd.DataFrame(
    df['Sequence'].apply(count_amino_acids).tolist(),
    columns=amino_acids,
    index=df.index
)
df = pd.concat([df, aa_counts], axis=1)



x = df[['length'] + amino_acids].to_numpy()
print(x.shape)

(1092, 21)


In [65]:
#Exercise 4
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size=0.3,      # 30% for testing
    random_state=42,    # for reproducibility
    stratify=y          # maintain class distribution
)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(764, 21)
(328, 21)
(764, 1)
(328, 1)


In [68]:
#Exercise 5
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='liblinear',max_iter=1000)
model.fit(X_train, y_train)


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(max_iter=1000, solver='liblinear')

In [70]:
#Exercise 6
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy*100}%")



Accuracy: 77.13414634146342%


Exercise 6

I got an accuracy of ~77%.

If I guessed the common class every time, the baseline would be around 58.7%.

My model appears to do better than the baseline.



Exercise 8

fit() is the learning step where the machine figures out the rules. predict() is the testing step where you feed the machine a new never seen before data set where it applies its learned rules. It then gives its predicted y values.

We split our train and test sets in order to know how well the model works on never before seen data.

Other features we could extract would be the number of hydrophobic or hydrophilic amino acids, pI and more.

PLM embeddings contain more information than the 21 number summary that I handcrafted.